sdk/core/Azure.Core/src/TelemetryDetails.cs

```diff
@@ -88,7 +86,14 @@ internal static string GenerateUserAgentString(Assembly clientAssembly, string?
                 version = version.Substring(0, hashSeparator);
             }
             runtimeInformation ??= new RuntimeInformationWrapper();
-            var platformInformation = EscapeProductInformation($"({runtimeInformation.FrameworkDescription}; {runtimeInformation.OSDescription})");
+
+            // RFC 9110 section 5.5 https://www.rfc-editor.org/rfc/rfc9110.txt#section-5.5 does not require any specific encoding : "Fields needing a greater range of characters
+            // can use an encoding, such as the one defined in RFC8187." RFC8187 is targeted at parameter values, almost always filename, so using url encoding here instead, which is
+            // more widely used. Since user-agent does not usually contain non-ascii, only encode when necessary.
+            // This was added to support operating systems with non-ascii characters in their release names.
+            var osDescription = ContainsNonAscii(runtimeInformation.OSDescription) ? WebUtility.UrlEncode(runtimeInformation.OSDescription) : runtimeInformation.OSDescription;
```

Comment:
Curious: would it make sense to fold this check into the escaping logic, which is already iterating over the characters of the string, or would that not catch all cases?

If it's equivalent from a perf perspective, then I like factoring it this way better per clarity.

Reply:
I could definitely fold this in but then wouldn't be able to use the .NET implementation for url encoding. I could bring over the url encode implementation and only return the encoded version if we saw a non-ascii character while encoding? Doing it in one step would improve the perf, but in most cases the encoding step won't be necessary so it would only improve for the non-ascii case.